libs/checkpoint/src/store/postgres.ts

```diff
@@ -0,0 +1,714 @@
+  value: Record<string, unknown>;
+  prefix: string;
+  created_at: Date;
+  updated_at: Date;
+  score?: number;
+}
+
+/**
+ * PostgreSQL-backed store with optional vector search using pgvector.
+ * 
+ * --- Basic setup and usage ---
+ * const store = new PostgresStore({
+ *   connectionString: "postgresql://user:pass@localhost:5432/dbname"
+ * });
+ * await store.setup(); // Run migrations once
+ * 
+ * --- Store and retrieve data ---
+ * await store.put(["users", "123"], "prefs", { theme: "dark" });
+ * const item = await store.get(["users", "123"], "prefs");
+ * 
+ * --- Vector search with embeddings ---
+ * const store = new PostgresStore({
+ *   connectionString: "postgresql://user:pass@localhost:5432/dbname",
+ *   index: {
+ *     dims: 1536,
+ *     embeddings: new OpenAIEmbeddings({ modelName: "text-embedding-3-small" }),
+ *     fields: ["text"] // specify which fields to embed
+ *   }
+ * });
+ * 
+ * --- Store documents ---
+ * await store.put(["docs"], "doc1", { text: "Python tutorial" });
+ * await store.put(["docs"], "doc2", { text: "TypeScript guide" });
+ * 
+ * --- Search by similarity ---
+ * const results = await store.search(["docs"], { query: "programming guides" });
+ */
+export class PostgresStore extends BaseStore {
+  private pool: InstanceType<typeof Pool>;
+
+  private indexConfig?: PostgresIndexConfig & {
+    __tokenizedFields?: Array<[string, string[]]>;
+  };
+
+  constructor(
+    config: PoolConfig & {
+      index?: PostgresIndexConfig;
+    }
+  ) {
+    super();
+    const { index, ...poolConfig } = config;
+    this.pool = new Pool(poolConfig);
+
+    if (index) {
+      this.indexConfig = {
+        ...index,
+        __tokenizedFields: (index.fields ?? ["$"]).map((p) => [
+          p,
+          p === "$" ? [p] : tokenizePath(p),
+        ]),
+      };
+    }
+  }
+
+  /**
+   * Set up the store database.
+   * Creates necessary tables and extensions if they don't exist.
+   */
+  async setup(): Promise<void> {
+    const client = await this.pool.connect();
+    try {
+      // Create migrations table
+      await client.query(`
+        CREATE TABLE IF NOT EXISTS store_migrations (
+          v INTEGER PRIMARY KEY
+        )
+      `);
+
+      // Get current version
+      const { rows: [row] } = await client.query(
+        "SELECT v FROM store_migrations ORDER BY v DESC LIMIT 1"
+      );
+      const version = row ? row.v : -1;
+
+      // Base migrations
+      const migrations = [
+        `
+        CREATE TABLE IF NOT EXISTS store (
+          prefix text NOT NULL,
+          key text NOT NULL,
+          value jsonb NOT NULL,
+          created_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP,
+          updated_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP,
+          PRIMARY KEY (prefix, key)
+        )
+        `,
+        `
+        CREATE INDEX CONCURRENTLY IF NOT EXISTS store_prefix_idx 
+        ON store USING btree (prefix text_pattern_ops)
+        `,
+      ];
```

Comment:
I'd recommend moving the migrations logic to its own file, as we'll likely add migrations over time if the scope of this tool changes. You can look to the `checkpoint-postgres` project to see how that was organized, if you want some inspiration. Alternatively, feel free to add a dependency on a decent migration management tool that works with postgres.